systemdspawner/systemdspawner.py

```diff
@@ -284,12 +284,11 @@ async def start(self):
             properties["PrivateDevices"] = "yes"
 
         if self.extra_paths:
-            env["PATH"] = "{extrapath}:{curpath}".format(
-                curpath=env["PATH"],
-                extrapath=":".join(
-                    [self._expand_user_vars(p) for p in self.extra_paths]
-                ),
-            )
+            new_path_list = [self._expand_user_vars(p) for p in self.extra_paths]
+            current_path = env.get("PATH")
```

Comment:
```suggestion
            current_path = env.get("PATH", os.defpath)
```

Reply:
This should be done without being conditional on `extra_paths` being set non-empty then, right?

Reply:
I've pushed [3d12e12](https://github.com/jupyterhub/systemdspawner/pull/144/commits/3d12e1213ce09cc8c8975ae0ee8c523ffa9d8e53) making us initialize PATH to os.defpath if unset, independently of if `extra_paths` has been set truthy or not.

Reply:
I think this should only be for reading PATH. This is the same effective value if PATH is omitted, so we shouldn't need to add it if we're not setting PATH. It's only relevant when _adding_ to PATH.

Reply:
It seems that by declaring env["PATH"] explicitly to `os.defpath` when no os.environ.get("PATH") is unset, we end up making this conditional truthy now but didn't before, and that makes our systemdspawner test suite fail with `jupyterhub-singleuser not found on /bin:/usr/bin`.

https://github.com/jupyterhub/systemdspawner/blob/45b06fefaf7751d2560da2bf60b9a34fbfe42f5a/systemdspawner/systemd.py#L135-L146

I think the situation is that systemdspawner about to spawn a server in JupyterHub 5.2 doesn't have PATH set, and when that happens, it declares PATH explicitly, and by doing that, it makes that the PATH in the systemd service started for the user, while the PATH for the user otherwise would be different - and still include jupyterhub-singleuser.

Overall I'm not sure whats the right call to do here at the moment.

Reply:
That maybe wasn't clear. I think this should only be for extra_paths when PATH is undefined. _if_ some systemd config has set the default path, that's still what should be used when PATH is unspecified, so PATH should remain unspecified. But extra_paths implies adding to the existing PATH, and this is the "default" default value for Python, which is probably but perhaps not necessarily the same as systemd.

Reply:
I missed the comment you made because GitHub UI didn't refresh properly =/

Okay deal! We'll then let this only influence the situation when `extra_paths` is set and we are reading PATH to prepend `extra_paths`!

Reply:
Thank you for helping me with this @minrk!!!

Reply:
Looking at the path resolution on L159, it uses `os.getenv(PATH, os.defpath)`, i.e. the Hub's PATH, but only for looking up jupyterhub-singleuser itself, and only if $PATH  is unspecified for the unit. it doesn't affect the default PATH of the process.

So systemdspawner will find JupyterHub-singleuser if:

- you specify PATH and jupyterhub-singleuser, or
- you don't specify PATH at all

But not if you specify PATH to something without singleuser, which seems inconsistent.

I think behavior might be more consistent if we either:

- never add Hub $PATH to the search for JupyterHub-singleuser (rely on correct PATH config of the systemd unit and/or absolute paths), or
- always do (singleuser should always be found, with current env as lowest priority fallback, no matter what unit $PATH is)

so behavior is more consistent across config situations.

We could also change the default for unit $PATH to hub $PATH, but this has higher risk of impacting real deployments. It should be quite rare and usually a mistake to leave $PATH undefined in a systemd unit, especially for user servers. Unless the launch command is an initialization script that is responsible for setting up $PATH.

Reply:
Btw, I think there's an opportunity to clear a few things up and be more consistent, but that's its own task, not needed here. The surprising test failures here just raise the issue.

Reply:
I've opened https://github.com/jupyterhub/systemdspawner/issues/146 to not loose track of this discussion.